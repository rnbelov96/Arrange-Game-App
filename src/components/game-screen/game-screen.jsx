import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { getCellsToClick, getDimention, getField } from '@/reducer/app/selectors';
import { AppActionCreators } from '@/reducer/app';

const Field = styled.div`
  display: flex;
  flex-wrap: wrap;
  @media (max-width: 1200px) {
    width: 50vw;
    height: 50vw;
  }
  @media (max-width: 991px) {
    width: 60vw;
    height: 60vw;
  }
  @media (max-width: 767px) {
    width: 70vw;
    height: 70vw;
  }
  @media (max-width: 575px) {
    width: 75vw;
    height: 75vw;
  }
  width: 45vw;
  height: 45vw;
`;

const Cell = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  border: 2px solid #bbb7b7;
  color: #bbb7b7;
  cursor: pointer;
  font-size: 35px;
  font-weight: bold;
  @media (max-width: 575px) {
    font-size: 30px;
  }
`;

const GameScreen = ({
  field, cellsToClick, onClick, dimention,
}) => {
  const cellsWidth = {
    three: '33%',
    four: '25%',
    five: '20%',
  };
  return (
    <Field>
      {field.map((el, index) => (
        <Cell
          key={`${index + 1}-cell`}
          style={{ width: `${cellsWidth[dimention]}` }}
          onClick={
            cellsToClick.includes(index) ? () => onClick(index) : null
          }
        >
          {el}
        </Cell>
      ))}
    </Field>
  );
};

GameScreen.propTypes = {
  field: PropTypes.arrayOf(PropTypes.number).isRequired,
  cellsToClick: PropTypes.arrayOf(PropTypes.number).isRequired,
  onClick: PropTypes.func.isRequired,
  dimention: PropTypes.string.isRequired,
};

const mapStateToProps = state => ({
  field: getField(state),
  cellsToClick: getCellsToClick(state),
  dimention: getDimention(state),
});

const mapDispatchToProps = dispatch => ({
  onClick: index => dispatch(AppActionCreators.moveCell(index)),
});

export default connect(mapStateToProps, mapDispatchToProps)(GameScreen);
