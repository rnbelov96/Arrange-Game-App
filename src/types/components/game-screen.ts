import { DimentionType } from '../general-types';

export type GameScreenPropsType = {
  field: (number|null)[],
  cellsToClick: number[],
  onClick: ((index: number) => void),
  dimention: DimentionType,
}
