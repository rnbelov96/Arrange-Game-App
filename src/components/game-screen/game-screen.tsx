import React from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';
import { Dispatch } from 'redux';
import { getCellsToClick, getDimention, getField } from '@/redux/app/selectors';
import { AppActionCreators } from '@/redux/app';
import { FullStateType } from '@/types/general-types';
import { AppActionType } from '@/types/redux/app-reducer';
import { GameScreenPropsType } from '@/types/components/game-screen';

const Field = styled.div`
  display: flex;
  flex-wrap: wrap;
  @media (max-width: 1200px) {
    width: 50vw;
    height: 50vw;
  }
  @media (max-width: 991px) {
    width: 60vw;
    height: 60vw;
  }
  @media (max-width: 767px) {
    width: 70vw;
    height: 70vw;
  }
  @media (max-width: 575px) {
    width: 75vw;
    height: 75vw;
  }
  width: 45vw;
  height: 45vw;
`;

const Cell = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  border: 2px solid #bbb7b7;
  color: #bbb7b7;
  cursor: pointer;
  font-size: 35px;
  font-weight: bold;
  @media (max-width: 575px) {
    font-size: 30px;
  }
`;

export const PureGameScreen: React.FunctionComponent<GameScreenPropsType> = ({
  field, cellsToClick, onClick, dimention,
}: GameScreenPropsType) => {
  const cellsWidth = {
    three: '33%',
    four: '25%',
    five: '20%',
  };
  return (
    <Field>
      {field.map((el, index) => (
        <Cell
          key={`${index + 1}-cell`}
          style={{ width: `${cellsWidth[dimention]}` }}
          onClick={
            cellsToClick.includes(index) ? () => onClick(index) : () => null
          }
        >
          {el}
        </Cell>
      ))}
    </Field>
  );
};
const mapStateToProps = (state: FullStateType) => ({
  field: getField(state),
  cellsToClick: getCellsToClick(state),
  dimention: getDimention(state),
});

const mapDispatchToProps = (dispatch: Dispatch<AppActionType>) => ({
  onClick: (index: number) => dispatch(AppActionCreators.moveCell(index)),
});

export const GameScreen = connect(mapStateToProps, mapDispatchToProps)(PureGameScreen);
