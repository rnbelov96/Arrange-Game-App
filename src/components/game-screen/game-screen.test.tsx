import React from 'react';
import renderer from 'react-test-renderer';
import { PureGameScreen } from './game-screen';

describe('Game Screen', () => {
  test('should render correctly with 3 dimention', () => {
    const tree = renderer
      .create(
        <PureGameScreen
          dimention="three"
          field={[7, null, 6, 8, 4, 1, 3, 5, 2]}
          cellsToClick={[0, 2, 4]}
          onClick={() => {}}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should render correctly with 4 dimention', () => {
    const tree = renderer
      .create(
        <PureGameScreen
          dimention="four"
          field={[12, 14, null, 6, 13, 7, 9, 8, 1, 2, 3, 15, 4, 10, 5, 11]}
          cellsToClick={[6, 3, 1]}
          onClick={() => {}}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should render correctly with 5 dimention', () => {
    const tree = renderer
      .create(
        <PureGameScreen
          dimention="five"
          field={[
            17,
            10,
            20,
            13,
            15,
            23,
            6,
            11,
            5,
            null,
            2,
            8,
            3,
            1,
            19,
            22,
            16,
            4,
            7,
            9,
            18,
            24,
            21,
            14,
            12,
          ]}
          cellsToClick={[14, 4, 8]}
          onClick={() => {}}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
