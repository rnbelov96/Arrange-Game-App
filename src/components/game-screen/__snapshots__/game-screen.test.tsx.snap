// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game Screen should render correctly with 3 dimention 1`] = `
<div
  className="sc-AykKC gEhdZJ"
>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "33%",
      }
    }
  >
    7
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "33%",
      }
    }
  />
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "33%",
      }
    }
  >
    6
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "33%",
      }
    }
  >
    8
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "33%",
      }
    }
  >
    4
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "33%",
      }
    }
  >
    1
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "33%",
      }
    }
  >
    3
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "33%",
      }
    }
  >
    5
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "33%",
      }
    }
  >
    2
  </div>
</div>
`;

exports[`Game Screen should render correctly with 4 dimention 1`] = `
<div
  className="sc-AykKC gEhdZJ"
>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "25%",
      }
    }
  >
    12
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "25%",
      }
    }
  >
    14
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "25%",
      }
    }
  />
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "25%",
      }
    }
  >
    6
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "25%",
      }
    }
  >
    13
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "25%",
      }
    }
  >
    7
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "25%",
      }
    }
  >
    9
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "25%",
      }
    }
  >
    8
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "25%",
      }
    }
  >
    1
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "25%",
      }
    }
  >
    2
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "25%",
      }
    }
  >
    3
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "25%",
      }
    }
  >
    15
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "25%",
      }
    }
  >
    4
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "25%",
      }
    }
  >
    10
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "25%",
      }
    }
  >
    5
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "25%",
      }
    }
  >
    11
  </div>
</div>
`;

exports[`Game Screen should render correctly with 5 dimention 1`] = `
<div
  className="sc-AykKC gEhdZJ"
>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "20%",
      }
    }
  >
    17
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "20%",
      }
    }
  >
    10
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "20%",
      }
    }
  >
    20
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "20%",
      }
    }
  >
    13
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "20%",
      }
    }
  >
    15
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "20%",
      }
    }
  >
    23
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "20%",
      }
    }
  >
    6
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "20%",
      }
    }
  >
    11
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "20%",
      }
    }
  >
    5
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "20%",
      }
    }
  />
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "20%",
      }
    }
  >
    2
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "20%",
      }
    }
  >
    8
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "20%",
      }
    }
  >
    3
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "20%",
      }
    }
  >
    1
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "20%",
      }
    }
  >
    19
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "20%",
      }
    }
  >
    22
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "20%",
      }
    }
  >
    16
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "20%",
      }
    }
  >
    4
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "20%",
      }
    }
  >
    7
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "20%",
      }
    }
  >
    9
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "20%",
      }
    }
  >
    18
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "20%",
      }
    }
  >
    24
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "20%",
      }
    }
  >
    21
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "20%",
      }
    }
  >
    14
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "20%",
      }
    }
  >
    12
  </div>
</div>
`;
